samples/siwx91x_ota/src/main.c

```diff
@@ -0,0 +1,751 @@
+		uint8_t ota_header_size = sizeof(sl_si91x_firmware_header_t);
+
+		status = sl_wifi_get_firmware_size((void *)ctx->flash_buffer, &ota_calc_image_size);
+		if (status != SL_STATUS_OK) {
+			printf("Unable to fetch firmware size. Status: 0x%x\n", status);
+			return -EIO;
+		}
+
+		/* Check if downloaded image size matches the calculated size */
+		if (ctx->ota_image_size != ota_calc_image_size) {
+			printf("Image size mismatch - expected: %u, calculated: %u (file may be "
+			       "corrupted)\n",
+			       ctx->ota_image_size, ota_calc_image_size);
+			return -EINVAL;
+		}
+
+		printf("Firmware size: %u bytes\n", ctx->ota_image_size);
+
+		/* Process firmware header */
+		status = sl_si91x_fwup_start(ctx->flash_buffer);
+		if (status != SL_STATUS_OK) {
+			printf("Failed to load firmware header (0x%x)\n", status);
+			return -EINVAL;
+		}
+
+		/* Load first chunk (after header) */
+		status = sl_si91x_fwup_load(ctx->flash_buffer + ota_header_size,
+					    ctx->flash_buffer_len - ota_header_size);
+		if (status != SL_STATUS_OK) {
+			return -EIO;
+		}
+	} else {
+		/* Handle subsequent packets */
+		status = sl_si91x_fwup_load(ctx->flash_buffer, ctx->flash_buffer_len);
+	}
+	printf("Fw load success for requested range=%d-%d\n", ctx->ota_range_start_byte,
+	       ctx->ota_range_end_byte);
+
+	/* Check if firmware update is completed */
+	if (status == SL_STATUS_SI91X_FW_UPDATE_DONE) {
+		return 1;
+	} else if (status != SL_STATUS_OK) {
+		return -EIO;
+	}
+	return 0;
+}
+
+/*
+ * @brief Handles retry logic for OTA operations
+ *
+ * @param ctx Pointer to application context
+ * @param operation_name Name of the operation being retried for logging
+ * @param cleanup_socket Whether to clean up socket resources before retrying
+ * @return true if retry should continue, false if max retries exceeded
+ */
+static bool ota_handle_retry(struct app_ctx *ctx, const char *operation_name, bool cleanup_socket)
+{
+	/* Clean up socket if needed */
+	if (cleanup_socket) {
+		ota_cleanup_resources(ctx);
+	}
+	if (++ctx->retry_count > 2) {
+		printf("Maximum retries exceeded, aborting OTA\n");
+		return false;
+	}
+	printf("Retrying %s (%d/%d)...\n", operation_name, ctx->retry_count, 2);
+	k_sleep(K_MSEC(1000));
+	return true;
+}
+
+/*
+ * @brief Main OTA application state machine
+ * @param ctx Pointer to application context
+ */
+static void ota_application_start(struct app_ctx *ctx)
+{
+	char *ssid = CONFIG_OTA_WIFI_SSID;
+	char *pwd = CONFIG_OTA_WIFI_PSK;
+	bool ota_complete = false;
+	const char *operation;
+	int ret = 0;
+
+	printf("OTA Application Started\n");
+	/* Main state machine loop */
+	while (ctx->retry_count < 3) {
+		switch (ctx->state) {
+		case OTA_STATE_SCAN:
+			/* Scan for networks and connect */
+			ota_wifi_start_scan(ctx);
+			ret = ota_wifi_connect(ssid, pwd, ctx);
+			if (ret) {
+				if (!ota_handle_retry(ctx, "Wi-Fi connection", false)) {
+					ctx->retry_count++;
+					continue;
+				}
+				break;
```

Comment:
I don't understand the complexity of this error management:
  - Here `break` and `continue`, will have the same behavior.
  - Why do you increment retry_count while retry_count is already incremented in `ota_handle_retry()`?

At the end, what is the difference with:

    if (ret) {
        ota_handle_retry(ctx, "Wi-Fi connection", false));
        continue;
    }